Issue 29: Deprecate --os

deadlysyn opened the pull request:
Deprecate --os in favor of --package-manager. Minor cleanup along the way.

This pull request makes the following changes:

  - Deprecate --os (shouldn't break builds and work mostly the same thanks to auto detection)
  - Add --package-manager
  - Minor updates to purl format for consistency across formats (<-- need feedback!)
  - Minor tweak to flag setup in chase.go inspired by iq.go (add consistency)
  - Update tests to use --package-manager

It relates to the following issue #s:

  - Addresses #29

cc @bhamail / @DarthHater / @zendern

zendern commented on packages/apk.go:
```diff
@@ -26,9 +26,9 @@ type Apk struct {
+func (a Apk) ExtractPurlsFromProjectList() (purls []string) {
 	for _, s := range a.ProjectList.Projects {
-		var purl = fmt.Sprintf("pkg:alpine/%s@%s", s.Name, s.Version)
+		var purl = fmt.Sprintf("pkg:apk/alpine/%s@%s", s.Name, s.Version)
```
Did we ever get confirmation that this is fine to change these paths??

I think if im reading this right. We were using the old "backward compatible" format and this meets the new format. 
https://ossindex.sonatype.org/doc/coordinates

deadlysyn replied:
I wanted to get feedback on which way we want to go with this. This was a "read by refactor" moment.  I followed the notes in the current spec and confirmed tests pass but not sure if we want to decouple this.

We also discussed debian being hard coded, so it might be worth fleshing out the format update + os auto detection in another PR. I'm not really sure if we want vulns reported specifically as "debian", "ubuntu", etc. (since there is lag between the various distro sources) or keep it generic since they're all "debs".

That's also true for fedora vs centos and other rhel clones...all rpm, but significant skew between when repos get updates. What's the right way to ensure each distro has the most accurate info about their specific vulns?

Would love input from the larger team, and if we want another PR I can use that to kick it off!

zendern commented on packages/yum.go:
```diff
@@ -26,9 +26,9 @@ type Yum struct {
+func (y Yum) ExtractPurlsFromProjectList() (purls []string) {
 	for _, s := range y.ProjectList.Projects {
-		var purl = fmt.Sprintf("pkg:rpm/%s@%s", s.Name, s.Version)
+		var purl = fmt.Sprintf("pkg:rpm/fedora/%s@%s", s.Name, s.Version)
```
Same statement here as the alpine one

deadlysyn replied:
Same as above...

zendern commented on cmd/chase.go:
```diff
@@ -225,10 +230,10 @@ func parseStdInList(list []string, operating *string) (packages.IPackage, error)
 		logLady.WithFields(logrus.Fields{
 			"list": list,
-		}).Trace("Chasing Fedora")
+		}).Trace("Chasing dnf")
```
In this case can we `log.Trace("Chasing <thing>")` just so we know for certain which ran

deadlysyn replied:
very nice that is so much cleaner :-)

zendern commented on cmd/chase.go:
```diff
@@ -240,7 +245,7 @@ func parseStdInList(list []string, operating *string) (packages.IPackage, error)
 		logLady.WithFields(logrus.Fields{
 			"list": list,
-		}).Trace("Chasing Yum")
+		}).Trace("Chasing yum")
```
Maybe for clarity while we are at it would you mind adding something like `(default case)` to the `Chasing yum` so that if looking at the debug log we know that we fell all the way through the switch

zendern commented on cmd/iq.go:
```diff
@@ -42,7 +42,8 @@ func init() {
 
 	pf := iqCmd.PersistentFlags()
-	pf.StringVar(&operating, "os", "", "Specify a value for the operating system type you want to scan (alpine, debian, fedora). Useful if autodetection fails and/or you want to explicitly set it.")
+	pf.StringVar(&packageManager, "os", "", "Specify a value for the operating system type you want to scan (alpine, debian, fedora). Useful if autodetection fails and/or you want to explicitly set it.")
```
So dumb question if we do this and set it to `&packageManager` wont that break everyone that is in theory currently passing in a real value here?? Or does Cobra just stop you since its been deprecated and spits out a message??

deadlysyn replied:
good question -- no it just throws the deprecation warning and continues.

i thought i'd tested this successfully, but now i'm not sure how we could get desired behavior without some more logic to map the os names to package manager names... i'm in the middle of distro hopping (perfect timing :-) ) so need to get docker going again and some VMs spun up to test this adequately. i'll hammer that out this weekend.

thanks for the feedback.

zendern commented:
@deadlysyn Overall looks good. Just a few minor questions/tweaks

deadlysyn commented:
To help move this forward, I've decoupled the purl format updates. Maybe we can tackle that as a separate PR. Since we're not passing os anymore, apt.go has to hard code Debian while both Alpine and Fedora don't include OS at all. It feels like we should be consistent one way or the other, but that doesn't need decided here.

To avoid strewing os references all over, I just updated the switch to support either --os or --package-manager strings. Seems to work locally, and also updated tests to cover both until the deprecated bits can be fully removed.

zendern commented:
> To help move this forward, I've decoupled the purl format updates. Maybe we can tackle that as a separate PR. Since we're not passing os anymore, apt.go has to hard code Debian while both Alpine and Fedora don't include OS at all. It feels like we should be consistent one way or the other, but that doesn't need decided here.
> 
> To avoid strewing os references all over, I just updated the switch to support either --os or --package-manager strings. Seems to work locally, and also updated tests to cover both until the deprecated bits can be fully removed.

Cool I opened #44 to deal with that then @deadlysyn .

bhamail reviewed (approved):
+1

I'm gonna merge this soon too - mostly because I created a merge conflict by doing a prior merge. I'll attempt to resolve the conflict first. 

BTW, please don't feel any need to wait on me to merge these. Once you have 'nuf approvals, go for it!
